### PR TITLE
combinators: avoid effectful parameters

### LINKED ISSUE #650: combinators: review methods with effectful parameters
Context: https://github.com/getkyo/kyo/pull/644#discussion_r1749344224

### PR DESCRIPTION
**Pull request by fwbrasil**
Fixes #650

Some of the methods I still kept receiving effectful parameters and I've also promoted `as` to the pending type since it's a common pattern.

### CODE REVIEW

File: kyo-prelude/shared/src/main/scala/kyo/kernel/Pending.scala [hunk 1]
```diff
@@ -45,6 +45,14 @@ object `<`:
         ): B < (S & S2) =
             map(_ => f)
 
+        inline def as[B, S2](other: B < S2)(
```
Comment (fwbrasil) [inline kyo-prelude/shared/src/main/scala/kyo/kernel/Pending.scala:48]: I noticed that the kernel still doesn't have scaladocs. I'll follow up separately.

File: kyo-combinators/shared/src/main/scala/kyo/Combinators.scala [hunk 2]
```diff
@@ -429,20 +405,19 @@ extension [A, S, E](effect: A < (Abort[Maybe.Empty] & S))
...
       */
-    def emptyAbortToFailure[S1](failure: => E < S1)(using f: Flat[A], frame: Frame): A < (S & S1 & Abort[E]) =
+    def emptyAbortToFailure(failure: E)(using Flat[A], Frame): A < (S & Abort[E]) =
         for
-            f   <- failure
             res <- effect.handleSomeAbort[Maybe.Empty]()
         yield res match
-            case Result.Fail(_)    => Abort.get(Result.Fail(f))
+            case Result.Fail(_)    => Abort.get(Result.Fail(failure))
             case Result.Panic(e)   => Abort.get(Result.Panic(e))
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Combinators.scala:421]: We should avoid re-allocating a `Panic`

File: kyo-combinators/shared/src/main/scala/kyo/Combinators.scala [hunk 3]
```diff
@@ -429,20 +405,19 @@ extension [A, S, E](effect: A < (Abort[Maybe.Empty] & S))
...
+    def emptyAbortToChoice(using Flat[A], Frame): A < (S & Choice) =
         effect.someAbortToChoice[Maybe.Empty]()
 
     /** Handles the Abort[Maybe.Empty] effect translating it to an Abort[E] effect.
       *
       * @return
       *   A computation that produces the result of this computation with the Abort[Maybe.Empty] effect translated to Abort[E]
       */
-    def emptyAbortToFailure[S1](failure: => E < S1)(using f: Flat[A], frame: Frame): A < (S & S1 & Abort[E]) =
+    def emptyAbortToFailure(failure: E)(using Flat[A], Frame): A < (S & Abort[E]) =
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Combinators.scala:416]: This needs to be by name.

File: kyo-combinators/shared/src/main/scala/kyo/Combinators.scala [hunk 4]
```diff
@@ -657,13 +631,9 @@ extension [A, S](effect: A < (S & Choice))
       * @return
       *   A computation that produces the result of this computation with Abort[E] effect
       */
-    def choiceToAbort[E, S1](error: => E < S1)(
-        using
-        Flat[A],
-        Frame
-    ): A < (S & S1 & Abort[E]) =
+    def choiceToAbort[E](error: E)(using Flat[A], Frame): A < (S & Abort[E]) =
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Combinators.scala:634]: error should be by name

pre-existing: combinator should return `Seq` or mention it's taking `head`

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 5]
```diff
@@ -29,7 +29,7 @@ extension (kyoObject: Kyo.type)
       * @return
       *   An effect that ensures the finalizer is executed when the effect is completed
       */
-    def addFinalizer(finalizer: => Unit < IO)(using Frame): Unit < (Resource & IO) =
+    def addFinalizer(finalizer: => Unit < Async)(using Frame): Unit < (Resource & IO) =
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:32]: Wouldn't this make more sense as an extension method?

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 6]
```diff
@@ -19,7 +19,7 @@ extension (kyoObject: Kyo.type)
       * @return
       *   An effect that manages the resource lifecycle using Resource and IO effects
       */
-    def acquireRelease[A, S](acquire: => A < S)(release: A => Unit < IO)(using Frame): A < (S & Resource & IO) =
+    def acquireRelease[A, S](acquire: => A < S)(release: A => Unit < Async)(using Frame): A < (S & Resource & IO) =
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:22]: https://github.com/getkyo/kyo/blob/main/kyo-core/shared/src/main/scala/kyo/Resource.scala#L54-L57

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 7]
```diff
@@ -111,12 +107,8 @@ extension (kyoObject: Kyo.type)
       * @return
       *   A new sequence with elements collected using the function
       */
-    def foreachPar[A, S, A1](
-        sequence: => Seq[A] < S
-    )(
-        useElement: A => A1 < Async
-    )(using Flat[A1], Frame): Seq[A1] < (S & Async) =
-        sequence.map(seq => Async.parallel(seq.map(useElement)))
+    def foreachPar[A, S, A1](sequence: Seq[A])(useElement: A => A1 < Async)(using Flat[A1], Frame): Seq[A1] < Async =
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:110]: Can this support Ctx?

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 8]
```diff
@@ -127,12 +119,8 @@ extension (kyoObject: Kyo.type)
...
       *   Discards the results of the function application and returns Unit
       */
-    def foreachParDiscard[A, S, Any](
-        sequence: => Seq[A] < S
-    )(
-        useElement: A => Any < Async
-    )(using Flat[Any], Frame): Unit < (S & Async) =
-        sequence.map(seq => Async.parallel(seq.map(v => useElement(v)))).unit
+    def foreachParDiscard[A, Any](sequence: Seq[A])(useElement: A => Any < Async)(using Flat[Any], Frame): Unit < Async =
+        Async.parallel(sequence.map(v => useElement(v))).unit
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:123]: Can this invoke `foreachPar`?

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 9]
```diff
@@ -243,11 +231,8 @@ extension (kyoObject: Kyo.type)
       * @return
       *   An effect that logs the message and error to the console
       */
-    inline def logInfo[S, S1](
-        message: => String < S,
-        err: => Throwable < S1
-    ): Unit < (S & S1 & IO) =
-        message.map(m => err.map(e => Log.info(m, e)))
+    inline def logInfo(message: String, err: Throwable): Unit < IO =
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:234]: I believe all Logging functions should accept functions by name.

File: kyo-combinators/shared/src/main/scala/kyo/Combinators.scala [hunk 10]
```diff
@@ -112,17 +100,11 @@ extension [A, S](effect: A < S)
...
-        Frame
-    ): A < (S & S1 & Async) =
-        limit.map { limit =>
-            Loop.indexed { i =>
-                if i >= limit then effect.map(Loop.done)
-                else effect.delayed(backoff(i)).as(Loop.continue)
-            }
+    def repeat(backoff: Int => Duration, limit: => Int)(using Flat[A], Frame): A < (S & Async) =
+        Loop.indexed { i =>
+            if i >= limit then effect.map(Loop.done)
```
[the remark this answers is not in the document]
Comment (fwbrasil) [reply kyo-combinators/shared/src/main/scala/kyo/Combinators.scala:105]: I've made it a by-value parameter instead

File: kyo-combinators/shared/src/main/scala/kyo/Combinators.scala (hunk 4 above)
Comment (fwbrasil) [reply kyo-combinators/shared/src/main/scala/kyo/Combinators.scala:634]: What do you think about removing these methods instead? It seems a foot gun even if we make scaladocs and naming more clear

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala (hunk 5 above)
Comment (fwbrasil) [reply kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:32]: I've added `ensuring` like ZIO

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala (hunk 7 above)
Comment (fwbrasil) [reply kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:110]: Added! It was also missing test coverage

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala (hunk 9 above)
Comment (fwbrasil) [reply kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:234]: They were missing `inline` as well

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 11]
```diff
@@ -89,8 +86,8 @@ extension (kyoObject: Kyo.type)
       * @return
       *   An effect that prints the message to the console
       */
-    def debugln[S](message: => String < S)(using Frame): Unit < (S & IO) =
-        message.map(m => Console.println(m))
+    def debugln[S](message: String)(using Frame): Unit < IO =
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:89]: This one seems less important to remove effectual parameter... Regardless there is a leftover `[S]`

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 12]
```diff
@@ -99,8 +96,8 @@ extension (kyoObject: Kyo.type)
       * @return
       *   An effect that fails with the given error
       */
-    def fail[E, S](error: => E < S)(using Frame): Nothing < (S & Abort[E]) =
-        error.map(e => Abort.fail(e))
+    def fail[E](error: E)(using Frame): Nothing < Abort[E] =
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:99]: This should be by name.

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 13]
```diff
@@ -111,12 +108,14 @@ extension (kyoObject: Kyo.type)
...
-    )(using Flat[A1], Frame): Seq[A1] < (S & Async) =
-        sequence.map(seq => Async.parallel(seq.map(useElement)))
+    def foreachPar[E, A, S, A1, Ctx](sequence: Seq[A])(useElement: A => A1 < (Abort[E] & Async))(
+        using
+        flat: Flat[A1],
+        boundary: Boundary[Ctx, Async],
+        reduce: Reducible[Abort[E]],
+        frame: Frame
+    ): Seq[A1] < (reduce.SReduced & Async & Ctx) =
+        Async.parallel[E, A1, Ctx](sequence.map(useElement))
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:118]:
```suggestion
    def foreachPar[E, A, S, A1, Ctx](sequence: Seq[A])(useElement: A => A1 < (Abort[E] & Async & Ctx))(
        using
        flat: Flat[A1],
        boundary: Boundary[Ctx, Async],
        reduce: Reducible[Abort[E]],
        frame: Frame
    ): Seq[A1] < (reduce.SReduced & Async & Ctx) =
        Async.parallel[E, A1, Ctx](sequence.map(useElement))
```

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 14]
```diff
@@ -127,12 +126,14 @@ extension (kyoObject: Kyo.type)
       * @return
       *   Discards the results of the function application and returns Unit
       */
-    def foreachParDiscard[A, S, Any](
-        sequence: => Seq[A] < S
-    )(
-        useElement: A => Any < Async
-    )(using Flat[Any], Frame): Unit < (S & Async) =
-        sequence.map(seq => Async.parallel(seq.map(v => useElement(v)))).unit
+    def foreachParDiscard[E, A, S, A1, Ctx](sequence: Seq[A])(useElement: A => A1 < (Abort[E] & Async))(
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:129]:
```suggestion
    def foreachParDiscard[E, A, S, A1, Ctx](sequence: Seq[A])(useElement: A => A1 < (Abort[E] & Async & Ctx))(
```

File: kyo-combinators/shared/src/main/scala/kyo/Combinators.scala (hunk 4 above)
Comment (fwbrasil) [reply kyo-combinators/shared/src/main/scala/kyo/Combinators.scala]: removed

File: kyo-combinators/shared/src/main/scala/kyo/Constructors.scala [hunk 15]
```diff
@@ -484,10 +462,9 @@ extension (kyoObject: Kyo.type)
...
-    def traverseParDiscard[A, S](
-        sequence: => Seq[A < Async] < S
-    )(using Flat[A < S], Frame): Unit < (S & Async) =
-        sequence.map(seq =>
-            foreachPar(seq.map(_.unit))(identity).unit
-        )
+    def traverseParDiscard[A](
+        sequence: => Seq[A < Async]
+    )(using Flat[A], Frame): Unit < Async =
+        foreachPar(sequence.map(_.unit))(identity).unit
```
Comment (hearnadam) [inline kyo-combinators/shared/src/main/scala/kyo/Constructors.scala:468]: Is the `.map(_.unit)` necessary?

**Review by hearnadam — APPROVED** [review]

**Pull request merged**